pkg/collectors/collectors.go

```diff
@@ -50,28 +55,116 @@ var (
 	invalidLabelCharRE = regexp.MustCompile(`[^a-zA-Z0-9_]`)
 )
 
-var AvailableCollectors = map[string]func(registry prometheus.Registerer, informerFactories []informers.SharedInformerFactory, opts *options.Options){
-	"cronjobs":                 RegisterCronJobCollector,
-	"daemonsets":               RegisterDaemonSetCollector,
-	"deployments":              RegisterDeploymentCollector,
-	"jobs":                     RegisterJobCollector,
-	"limitranges":              RegisterLimitRangeCollector,
-	"nodes":                    RegisterNodeCollector,
-	"pods":                     RegisterPodCollector,
-	"replicasets":              RegisterReplicaSetCollector,
-	"replicationcontrollers":   RegisterReplicationControllerCollector,
-	"resourcequotas":           RegisterResourceQuotaCollector,
-	"services":                 RegisterServiceCollector,
-	"statefulsets":             RegisterStatefulSetCollector,
-	"persistentvolumes":        RegisterPersistentVolumeCollector,
-	"persistentvolumeclaims":   RegisterPersistentVolumeClaimCollector,
-	"namespaces":               RegisterNamespaceCollector,
-	"horizontalpodautoscalers": RegisterHorizontalPodAutoScalerCollector,
-	"endpoints":                RegisterEndpointCollector,
-	"secrets":                  RegisterSecretCollector,
-	"configmaps":               RegisterConfigMapCollector,
+// RegisterCollectors initializes and registers collectors.
+func CreateCollectors(
+	kubeClient clientset.Interface,
+	enabledCollectors options.CollectorSet,
+	namespaces options.NamespaceList,
+	opts *options.Options,
+) ([]Collector, error) {
+
+	collectorFactory := newCollectorFactory(kubeClient, opts)
+	collectors := []Collector{}
+
+	activeCollectorNames := []string{}
+	for c := range enabledCollectors {
+		constructor, ok := availableCollectors[c]
+		if ok {
+			collector, err := collectorFactory.create(constructor)
+			if err != nil {
+				return nil, err
+			}
+			activeCollectorNames = append(activeCollectorNames, c)
+			collectors = append(collectors, collector)
+		}
+	}
+
+	glog.Infof("Active collectors: %s", strings.Join(activeCollectorNames, ","))
+
+	return collectors, nil
+}
+
+type Collector interface {
+	createListWatch(clientset.Interface) cache.ListWatch
+	generateMetrics(obj interface{}) []*metrics.Metric
+	expectedType() interface{}
+	setStore(*metricsstore.MetricsStore)
+	Collect() []*metrics.Metric
+}
```

Comment:
nit: this public interface with both public and private methods looks iffy. Do you really need it that way?